// @udecode/plate-alignment/dist/index.es.js
import { createPluginFactory, getPluginType, ELEMENT_DEFAULT, mapInjectPropsToPlugin, usePlateEditorState, isCollapsed, findNode, isDefined, usePlateEditorRef, focusEditor, getPluginInjectProps, unsetNodes, setElements, isBlock } from '@udecode/plate-common';

const KEY_ALIGN = 'align';

/**
 * Creates a plugin that adds alignment functionality to the editor.
 */
const createAlignPlugin = createPluginFactory({
  key: KEY_ALIGN,
  then: editor => ({
    inject: {
      props: {
        nodeKey: KEY_ALIGN,
        defaultNodeValue: 'left',
        styleKey: 'textAlign',
        validNodeValues: ['left', 'center', 'right', 'justify'],
        validTypes: [getPluginType(editor, ELEMENT_DEFAULT)]
      }
    },
    then: (_, plugin) => mapInjectPropsToPlugin(editor, plugin, {
      deserializeHtml: {
        getNode: (el, node) => {
          if (el.style.textAlign) {
            node[plugin.key] = el.style.textAlign;
          }
        }
      }
    })
  })
});

const useAlignDropdownMenuState = () => {
  const editor = usePlateEditorState();
  let value = 'left';
  if (isCollapsed(editor === null || editor === void 0 ? void 0 : editor.selection)) {
    const entry = findNode(editor, {
      match: n => isDefined(n[KEY_ALIGN])
    });
    if (entry) {
      const nodeValue = entry[0][KEY_ALIGN];
      if (nodeValue === 'right') value = 'right';
      if (nodeValue === 'center') value = 'center';
      if (nodeValue === 'justify') value = 'justify';
    }
  }
  return {
    value
  };
};
const useAlignDropdownMenu = ({
  value
}) => {
  const editor = usePlateEditorRef();
  return {
    radioGroupProps: {
      value,
      onValueChange: newValue => {
        setAlign(editor, {
          value: newValue,
          key: KEY_ALIGN
        });
        focusEditor(editor);
      }
    }
  };
};

const setAlign = (editor, {
  key = KEY_ALIGN,
  value,
  setNodesOptions
}) => {
  const {
    validTypes,
    defaultNodeValue,
    nodeKey
  } = getPluginInjectProps(editor, key);
  const match = n => {
    return isBlock(editor, n) && !!validTypes && validTypes.includes(n.type);
  };
  if (value === defaultNodeValue) {
    unsetNodes(editor, nodeKey, {
      match,
      ...setNodesOptions
    });
  } else {
    setElements(editor, {
      [nodeKey]: value
    }, {
      match: match,
      ...setNodesOptions
    });
  }
};

export { KEY_ALIGN, createAlignPlugin, setAlign, useAlignDropdownMenu, useAlignDropdownMenuState };
