// @udecode/plate-alignment/dist/index.js
'use strict';

var plateCommon = require('@udecode/plate-common');

const KEY_ALIGN = 'align';

/**
 * Creates a plugin that adds alignment functionality to the editor.
 */
const createAlignPlugin = plateCommon.createPluginFactory({
  key: KEY_ALIGN,
  then: editor => ({
    inject: {
      props: {
        nodeKey: KEY_ALIGN,
        defaultNodeValue: 'left',
        styleKey: 'textAlign',
        validNodeValues: ['left', 'center', 'right', 'justify'],
        validTypes: [plateCommon.getPluginType(editor, plateCommon.ELEMENT_DEFAULT)]
      }
    },
    then: (_, plugin) => plateCommon.mapInjectPropsToPlugin(editor, plugin, {
      deserializeHtml: {
        getNode: (el, node) => {
          if (el.style.textAlign) {
            node[plugin.key] = el.style.textAlign;
          }
        }
      }
    })
  })
});

const useAlignDropdownMenuState = () => {
  const editor = plateCommon.usePlateEditorState();
  let value = 'left';
  if (plateCommon.isCollapsed(editor === null || editor === void 0 ? void 0 : editor.selection)) {
    const entry = plateCommon.findNode(editor, {
      match: n => plateCommon.isDefined(n[KEY_ALIGN])
    });
    if (entry) {
      const nodeValue = entry[0][KEY_ALIGN];
      if (nodeValue === 'right') value = 'right';
      if (nodeValue === 'center') value = 'center';
      if (nodeValue === 'justify') value = 'justify';
    }
  }
  return {
    value
  };
};
const useAlignDropdownMenu = ({
  value
}) => {
  const editor = plateCommon.usePlateEditorRef();
  return {
    radioGroupProps: {
      value,
      onValueChange: newValue => {
        setAlign(editor, {
          value: newValue,
          key: KEY_ALIGN
        });
        plateCommon.focusEditor(editor);
      }
    }
  };
};

const setAlign = (editor, {
  key = KEY_ALIGN,
  value,
  setNodesOptions
}) => {
  const {
    validTypes,
    defaultNodeValue,
    nodeKey
  } = plateCommon.getPluginInjectProps(editor, key);
  const match = n => {
    return plateCommon.isBlock(editor, n) && !!validTypes && validTypes.includes(n.type);
  };
  if (value === defaultNodeValue) {
    plateCommon.unsetNodes(editor, nodeKey, {
      match,
      ...setNodesOptions
    });
  } else {
    plateCommon.setElements(editor, {
      [nodeKey]: value
    }, {
      match: match,
      ...setNodesOptions
    });
  }
};

exports.KEY_ALIGN = KEY_ALIGN;
exports.createAlignPlugin = createAlignPlugin;
exports.setAlign = setAlign;
exports.useAlignDropdownMenu = useAlignDropdownMenu;
exports.useAlignDropdownMenuState = useAlignDropdownMenuState;
